src/task.ts

```diff
@@ -0,0 +1,309 @@
+  const editor = vscode.window.activeTextEditor;
+  if (!editor) {
+    vscode.window.showErrorMessage('No active editor found.');
+    return;
+  }
+  const document = editor.document;
+  if (document.languageId !== 'chapel') {
+    vscode.window.showErrorMessage('The current file is not a Chapel file.');
+    return;
+  }
+  const wsPath = getWorkspaceFolder();
+  const filePath = path.resolve(document.fileName);
+  const tasks =
+    (await Promise.all([vscode.tasks.fetchTasks({
+      type: ChplBuildType,
+    }), vscode.tasks.fetchTasks({
+      type: ChplRunType,
+    })])).flat()
+      .filter(task => {
+        const definition = task.definition as ChplBuildTaskDefinition | ChplRunTaskDefinition;
+
+        const rootFile = path.resolve(resolveWorkspaceFolder(wsPath, definition.rootFile || ''));
+        return rootFile === filePath;
+      });
+
+  if (tasks.length === 0) {
+    vscode.window.showErrorMessage('No tasks found for this file.');
+    return;
+  }
+  const selectedTask = await vscode.window.showQuickPick(tasks.map(task => task.name), {
+    placeHolder: 'Select a task to run',
+  });
+  if (!selectedTask) {
+    return;
+  }
+  const task = tasks.find(t => t.name === selectedTask);
+  if (!task) {
+    vscode.window.showErrorMessage('Task not found.');
+    return;
+  }
+  await vscode.tasks.executeTask(task);
+
+
+}
+export function registerChapelTaskProvider(context: vscode.ExtensionContext): void {
+  context.subscriptions.push(vscode.tasks.registerTaskProvider(
+    ChplRunType,
+    new ChplRunTaskProvider()
+  ));
+  context.subscriptions.push(vscode.tasks.registerTaskProvider(
+    ChplBuildType,
+    new ChplBuildTaskProvider()
+  ));
+
+  context.subscriptions.push(
+    vscode.commands.registerCommand('chapel.runFileTask', chpl_runFileTask));
+}
+
+
+
+function getEnv(def: ChplBuildTaskDefinition | ChplRunTaskDefinition | undefined = undefined): Map<string, string> {
+  let envMap = def?.env ? new Map<string, string>(Object.entries(def.env)) : new Map<string, string>();
+  const baseEnv = getEnvAffectingChapel(["PATH"]);
+  // merge the base env with the user provided env, prefer user provided env
+  for (const [key, value] of baseEnv) {
+    if (!envMap.has(key)) {
+      envMap.set(key, value);
+    }
+  }
+  return envMap;
+}
+
+function replaceWithWorkspaceFolder(wsPath: string | undefined, filePath: string): string {
+  if (wsPath) {
+    if (filePath.startsWith(`${wsPath}/`)) {
+      return filePath.replace(wsPath, "${workspaceFolder}");
+    } else if (filePath === wsPath) {
+      return "${workspaceFolder}";
+    }
+  }
+  return filePath;
+}
+function resolveWorkspaceFolder(wsPath: string | undefined, filePath: string): string {
+  if (wsPath && filePath.includes("${workspaceFolder}")) {
+    return filePath.replace("${workspaceFolder}", wsPath);
+  }
+  return filePath;
+}
+
+function getWorkspaceFolder(): string | undefined {
+  return vscode.workspace.workspaceFolders?.[0]?.uri.fsPath;
```

Comment:
VS Code has the ability to have a workspace with many root level folders. This assumes that the correct one would always be the first root. 

Although I'm not sure what we could do to remedy this other than ask the user to specify the workspace folder.

Reply:
This function effectively returns the result of `${workspaceFolder}`.

Per [the vscode docs](https://code.visualstudio.com/docs/reference/variables-reference#_variables-scoped-per-workspace-folder), it should be resolved based on the folder where its used. That case isn't really handled here and I don't have a good way to remedy that

Reply:
Okay, that sounds good to me. We should record this in an issue or at least a comment in the code.

Reply:
I've made a note about this in my private task tracking, but I'll also add a TODO comment here